examples: add blecent and bleprph similar to mynewt-nimble apps

maxikrie opened the pull request:
## Summary
These example apps create a BLE central and peripheral based on mynewt-nimBLE and work in companion. The apps are mostly based on their equivalents, which can be found at https://github.com/apache/mynewt-nimble/tree/master/apps

## Impact

## Testing
These were successfully tested on two nrf52840-dk boards.

cederom reviewed (changes requested):
Thank you @maxikrie :-) Please fix the code formatting issues: https://github.com/apache/nuttx-apps/actions/runs/11602988992/job/32309359887?pr=2807 :-)

acassis commented on examples/nimble_blecent/misc.c:
```diff
@@ -0,0 +1,210 @@
+/**
+ * Utility function to log an array of bytes.
+ */
+void
+print_bytes(const uint8_t *bytes, int len)
```
@maxikrie I think it can fit in a single line

raiden00pl commented:
why not modify the current nimble peripheral example (https://github.com/apache/nuttx-apps/blob/master/examples/nimble/nimble_main.c)? do we need 2 different examples for this?

maxikrie commented on examples/nimble_blecent/misc.c:
```diff
@@ -0,0 +1,210 @@
+    int colon;
+
+    colon = 0;
+    while (om != NULL) {
+        if (colon) {
+            printf(":");
+        } else {
+            colon = 1;
+        }
+        print_bytes(om->om_data, om->om_len);
+        om = SLIST_NEXT(om, om_next);
+    }
```
I will make changes and adhere to the standard. I am sorry that I didn't follow it immediately, it has been a while since my last pull request and I didn't see the documentation on contributing beforehand.

raiden00pl commented:
@maxikrie Mynewt The mynewt coding standard is far from the NuttX coding standard, so the code will be a lot different anyway. Keeping 2 examples demonstrating the same functionality seems pointless to me. The easiest thing to do would be to change the name of existing `examples/nimble` to `examples/nimble_bleprph` (or `nimble_prph`) and add all your changes there. 
Or you can delete `examples/nimble` but then you will have more work to adapt the Mynewt code to the NuttX coding standard.

raiden00pl commented on examples/nimble_blecent/Kconfig:
```diff
@@ -0,0 +1,30 @@
+# see the file kconfig-language.txt in the NuttX tools repository.
+#
+
+config EXAMPLES_NIMBLE_BLECENT
```
what do you think about using the names `nimble_cent` and `nimble_prph` ?

maxikrie replied:
Fine with me, I just went along with as the examples are named in nimble.

maxikrie replied:
Also, I vote for deleting the legacy examples/nimble.

cederom commented:
CI checkpatch is now fixed, CI restarted :-)

cederom commented:
Source code still needs reformat, @maxikrie  please take a look at:
* https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md
* https://nuttx.apache.org/docs/latest/contributing/coding_style.html

maxikrie commented:
@cederom Is there a way to check whether my changes will pass the check before committing? I read the documentation on [pre-commi](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#using-pre-commit), but it seems to that this only applies to the nuttx repository - at least this repository is missing the required [.pre-commit-config.yaml](https://github.com/apache/nuttx/blob/master/.pre-commit-config.yaml).

I have installed pre-commit in the nuttx repository and tested it on branch nuttx-12.7.0, where everything is fine. I then naively tried from within the nuttx repository:
`pre-commit run --files ../apps/*`
which gives all sorts of strange errors.

xiaoxiang781216 commented:
> @cederom Is there a way to check whether my changes will pass the check before committing? I read the documentation on [pre-commit](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#using-pre-commit), but it seems that this only applies to the nuttx repository - at least this repository is missing the required [.pre-commit-config.yaml](https://github.com/apache/nuttx/blob/master/.pre-commit-config.yaml).
> 
> I have installed pre-commit in the nuttx repository and tested it on branch nuttx-12.7.0, where everything is fine. I then naively tried from within the nuttx repository: `pre-commit run --files ../apps/*` which gives all sorts of strange errors.

please run the following command to check the style:
../nuttx/tools/checkpatch.sh -g HEAD~...HEAD

acassis commented:
ping @maxikrie

maxikrie commented:
I fixed the formatting issues and fixed an error in the Makefile (I previously had only tested the CMake build), but there are again failed checks. Can somebody check? I am not sure what to make of them.. @xiaoxiang781216 @cederom

cederom commented:
Thank you @maxikrie :-)

Most of the formatting issues detected by CI are in the codebase already:
https://github.com/apache/nuttx-apps/actions/runs/12679172534/job/35340025084?pr=2807

Looks like these still needs an update?
* the ones with comment.
* builtin/builtin_list.c.
* canutils/candump/candump.c

maxikrie commented:
I am not sure, should I look into this? I am just checking since I didn't touch these and they didn't appear to be an issue when I run the nxstyle tool. I want to make sure this is not an issue with the CI.

raiden00pl commented on examples/nimble_bleprph/nimble_bleprph_main.c:
```diff
@@ -0,0 +1,447 @@
+ * The application associates a GAP event callback with each connection
+ * that forms. bleprph uses the same callback for all connections.
+ *
+ * @param event                 The type of event being signalled.
+ * @param ctxt                  Various information pertaining to the event.
+ * @param arg                   Application-specified argument; unuesd by
+ *                                  bleprph.
+ *
+ * @return                      0 if the application successfully handled the
+ *                                  event; nonzero on failure.  The semantics
+ *                                  of the return code is specific to the
+ *                                  particular GAP event being signalled.
+ */
```
doxygen is not allowed. The correct comment block for function looks like this:

```
/****************************************************************************
 * Private Functions
 ****************************************************************************/

/****************************************************************************
 * Name: <Static function name>
 *
 * Description:
 *   Description of the operation of the static function.
 *
 * Input Parameters:
 *   A list of input parameters, one-per-line, appears here along with a
 *   description of each input parameter.
 *
 * Returned Value:
 *   Description of the value returned by this function (if any),
 *   including an enumeration of all possible error values.
 *
 * Assumptions/Limitations:
 *   Anything else that one might need to know to use this function.
 *
 ****************************************************************************/
 ```

raiden00pl commented:
does anyone have any idea why CI is returning nxstyle errors unrelated to PR content?

<img width="735" alt="image" src="https://github.com/user-attachments/assets/5012fd41-a6ec-4836-8ebd-1ba758744002" />
What is this `Unchanged files with check annotations Preview` ? any ideas @lupyuen @simbit18 @xiaoxiang781216 ?

cederom commented:
ping @maxikrie :-)

maxikrie commented on examples/nimble_blecent/misc.c:
```diff
@@ -0,0 +1,243 @@
+
+/**
+ * Utility function to log an array of bytes.
+ */
```
Done

maxikrie commented on examples/nimble_blecent/misc.c:
```diff
@@ -0,0 +1,243 @@
+  return buf;
+}
+
+void print_uuid(const ble_uuid_t *uuid)
```
Done

maxikrie commented on examples/nimble_blecent/misc.c:
```diff
@@ -0,0 +1,210 @@
+}
+
+char *
+addr_str(const void *addr)
```
Done

xiaoxiang781216 commented:
@maxikrie please sqaush your patch into by git rebase --interactive HEAD~8

jerpelea reviewed (commented):
should be ok after squash

cederom commented:
Thank you @maxikrie, awesome work!! :-) Please squash so we have everything bundled into one single commit with common description and we are ready to go :-)

cederom reviewed (changes requested):
Thank you for the squash @maxikrie just `git commit -s` missing and we are ready to go :-)

cederom reviewed (approved):
Perfect! Thank you @maxikrie :-)

maxikrie commented:
@cederom I am confused why there are again long line issues? I fixed them previously and also  ../nuttx/tools/checkpatch.sh -g HEAD~...HEAD doesn't complain..

acassis commented:
> @cederom I am confused why there are again long line issues? I fixed them previously and also ../nuttx/tools/checkpatch.sh -g HEAD~...HEAD doesn't complain..

Please run checkpatch.sh to confirm that everything was fixed, CI is reporting long lines:

../nuttx/tools/checkpatch.sh -m -g 7f424c3e8ddaa0a6a828175b8e4205c1cc0a2a8a..HEAD
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:51:79: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:52:81: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:296:78: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:311:80: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:596:79: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:620:79: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_blecent/peer.c:621:82: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_bleprph/gatt_svr.c:108:78: error: Long line found
Error: /home/runner/work/nuttx-apps/nuttx-apps/apps/examples/nimble_bleprph/gatt_svr.c:185:81: error: Long line found
Error: Process completed with exit code 1.

cederom commented on examples/nimble_blecent/peer.c:
```diff
@@ -0,0 +1,926 @@
+int peer_svc_is_empty(const struct peer_svc *svc);
+
+uint16_t
+chr_end_handle(FAR const struct peer_svc *svc, FAR const struct peer_chr *chr);
```
Yeah, line is too long.. please take a look at https://nuttx.apache.org/docs/latest/contributing/coding_style.html :-)

cederom reviewed (changes requested):
Sorry, still formatting issues, please take a look at https://nuttx.apache.org/docs/latest/contributing/coding_style.html how to fix. Thanks! :-)

cederom commented:
Thank you @maxikrie :-) Looks almost done... just a line to be removed from a git commit message after signature (`git commit --amend`).. and we are ready to go :-) THANKS!! :-)
```

examples: add blecent and bleprph similar to mynewt-nimble apps

These example apps create a BLE central and peripheral based on
mynewt-nimBLE and work in companion. The apps are mostly based
on their equivalents, which can be found at
https://github.com/apache/mynewt-nimble/tree/master/apps

Fix formatting

Fix Makefiles

Format function parameters

Remove Doxygen comments

Add more function names and bug fix

Signed-off-by: Max Kriegleder <max.kriegleder@gmail.com>

Fix formatting
```

jerpelea reviewed (changes requested):
please add before the license text 

 * SPDX-License-Identifier: Apache-2.0
 *

cederom reviewed (approved):
Thank you @maxikrie AMAZING WORK!! :-)

maxikrie commented:
> * SPDX-License-Identifier: Apache-2.0

Where should this go? In every file?

xiaoxiang781216 commented:
> > @jerpelea
> > > * SPDX-License-Identifier: Apache-2.0
> > 
> > 
> > Where should this go? In every file?
> 
> every file that had a license ex: https://github.com/apache/nuttx/blob/master/boards/arm64/imx8/imx8qm-mek/src/imx8_appinit.c

@jerpelea please provide the mark with new patch.
